asterius/test/ghc-testsuite.hs

```diff
@@ -37,27 +64,169 @@ getTestCases = do
         readFileNullable (c -<.> "stdout") <*>
         readFileNullable (c -<.> "stderr")
 
-runTestCase :: TestCase -> IO (LBS.ByteString, LBS.ByteString)
+
+
+
+data TestOutcome = TestSuccess | TestFailure deriving(Eq, Show, Generic)
+
+instance ToJSON TestOutcome where
+
+
+data TestRecord = TestRecord
+  { trOutcome :: !TestOutcome
+  , trPath :: !FilePath -- ^ Path of the test case
+  , trErrorMessage :: !String -- ^ If the test failed, then the error message associated to the failure.
+  } deriving(Generic)
+
+instance ToJSON TestRecord where
+
+
+
+
+-- | Log of tests that have run
+newtype TestLog = TestLog [TestRecord] deriving(Semigroup, Monoid, Generic)
+
+instance ToJSON TestLog where
+
+-- | Append a value to the test log in safe way when we have multiple threads
+consTestLog :: TestRecord -> IORef TestLog -> IO ()
+consTestLog tr tlref = modifyIORef' tlref (\(TestLog tl) -> TestLog $ tr:tl)
+
+
+-- | What happened when you tried to run the test
+data RunOutcome = RunSuccess | RunFailure String deriving(Eq)
+
+
+-- [Note: RunOutcome's Show instance]
+-- | Have the Show instance print the exception after the separator  so we can
+-- | strip out the separator in the printer
+-- | This way, our custom ingredient can still serialize all the information
+-- | that comes after the ``, but when we print, we strip out the leading `:`
+-- | and all text that follows it.
+separator :: Char
+separator = 'γ'
+
+instance Show RunOutcome where
+  show (RunSuccess) = "RunSuccess"
+  show (RunFailure e) = "RunFailure" <> [separator] <> (show e)
+
+instance ToJSON RunOutcome where
+    toJSON RunSuccess = toJSON . show $ RunSuccess
+    toJSON (RunFailure f) = toJSON $ "RunFailure(" <> show f <> ")"
+
+runTestCase :: TestCase -> IO ()
 runTestCase TestCase {..} = do
-  _ <- readProcess "ahc-link" ["--input-hs", casePath] ""
+  _ <- readProcess "ahc-link" ["--input-hs", casePath, "--binaryen"] ""
   mod_buf <- LBS.readFile $ casePath -<.> "wasm"
   withJSSession defJSSessionOpts $ \s -> do
     i <- newAsteriusInstance s (casePath -<.> "lib.mjs") mod_buf
     hsInit s i
-    hsMain s i
+    -- | Try to run main. If we throw an exception, return a
+    -- RunFailure with the error message.
+    ro <- (hsMain s i *> pure RunSuccess)
+      `catch` (\(e :: SomeException) -> pure . RunFailure. show $ e)
+    -- | Check that the run succeeded. If it did not, report a failing
+    -- test case
+    ro `shouldBe` RunSuccess
+
+    -- | If the run succeded, now compare outputs.
     hs_stdout <- hsStdOut s i
     hs_stderr <- hsStdErr s i
-    pure (hs_stdout, hs_stderr)
 
+    hs_stdout `shouldBe` caseStdOut
+    hs_stderr `shouldBe` caseStdErr
+
+
+-- | @cheng: Why is this called `makeTestTree`? should it not be called
```

Comment:
Because `makeTestTree` produces a `TestTree`, but doesn't really run the test by the time the result is returned in `IO`.

Reply:
Indeed, realised this later. Will kill the comment, thanks!

Reply:
I think we should rename `runTestCase` to `makeTestCase`? because it too produces what the test case should be...